File: lab_1/machine_constants.cpp
#include <iostream>
#include <limits>

int main() {
    int k = 0;
    float num_a = 1;
    double num_b = 1;
    long double num_c = 1;

    std::cout << "------------------------------------" << std::endl;
    while (num_a != 0) {
        num_a /= 2;
        k++;
    }
    std::cout << "float machine zero = 2^(" << -k << ")" << std::endl;

    k = 0;
    num_a = 1;
    while (num_a < std::numeric_limits<float>::max()) {
        num_a *= 2;
        k++;
    }
    std::cout << "float machine infinity = 2^" << k << std::endl;

    k = 0;
    num_a = 1;
    while (1 + num_a > 1) {
        num_a /= 2;
        k++;
    }
    std::cout << "float machine epsilon = 2^(" << -k << ")" << std::endl;
    std::cout << "------------------------------------" << std::endl;
    while (num_b != 0) {
        num_b /= 2;
        k++;
    }
    std::cout << "double machine zero = 2^(" << -k << ")" << std::endl;

    k = 0;
    num_b = 1;
    while (num_b < std::numeric_limits<double>::max()) {
        num_b *= 2;
        k++;
    }
    std::cout << "double machine infinity = 2^" << k  << std::endl;

    k = 0;
    num_b = 1;
    while (1 + num_b > 1) {
        num_b /= 2;
        k++;
    }
    std::cout << "double machine epsilon = 2^(" << -k << ")" << std::endl;
    std::cout << "------------------------------------" << std::endl;
    while (num_c != 0) {
        num_c /= 2;
        k++;
    }
    std::cout << "long double machine zero = 2^(" << -k << ")" << std::endl;

    k = 0;
    num_c = 1;
    while (num_c < std::numeric_limits<long double>::max()) {
        num_c *= 2;
        k++;
    }
    std::cout << "long double machine infinity = 2^" << k << std::endl;

    k = 0;
    num_c = 1;
    while (1 + num_c > 1) {
        num_c /= 2;
        k++;
    }
    std::cout << "long double machine epsilon = 2^(" << -k << ")" << std::endl;



    return 0;
}

File: lab_1/machine_constants.py
import numpy as np
import warnings

type_display = {
    'float16': np.float16,
    'float32': np.float32,
    'float64': np.float64
}


def calculate_machine_constants(type_name):
    k = 0
    number_type = type_display[type_name]
    num = number_type(1)
    while num != 0:
        num = number_type(num / 2)
        k += 1
    print(f"{type_name} machine zero: 2^({-k})")

    k = 0
    num = number_type(1)
    while num != np.inf:
        num = number_type(num * 2)
        k += 1
    print(f"{type_name} machine infinity: 2^{k}")

    k = 0
    num = number_type(1)
    while 1.0 + num > 1.0:
        num = number_type(num / 2)
        k += 1
    print(f"{type_name} machine epsilon: 2^({-k})")


with warnings.catch_warnings():
    warnings.simplefilter("ignore")
    for typename in type_display.keys():
        print()
        calculate_machine_constants(typename)
